cpp/modmesh/buffer/BufferExpander.hpp

```diff
@@ -0,0 +1,173 @@
+
+#include <stdexcept>
+#include <memory>
+#include <algorithm>
+#include <sstream>
+
+namespace modmesh
+{
+
+/**
+ * Untyped and growing memory buffer for contiguous data storage.  The internal
+ * expandable memory buffer cannot be used externally.
+ */
+class BufferExpander
+    : public std::enable_shared_from_this<BufferExpander>
+{
+
+private:
+
+    struct ctor_passkey
+    {
+    };
+
+public:
+
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+
+    template <typename... Args>
+    static std::shared_ptr<BufferExpander> construct(Args &&... args)
+    {
+        return std::make_shared<BufferExpander>(std::forward<Args>(args)..., ctor_passkey());
+    }
+
+    std::shared_ptr<BufferExpander> clone()
+    {
+        return BufferExpander::construct(copy_concrete(), /*clone*/ false);
+    }
+
+    BufferExpander(std::shared_ptr<ConcreteBuffer> const & buf, bool clone, ctor_passkey const &)
+        : m_concrete_buffer(clone ? buf->clone() : buf)
+        , m_begin(m_concrete_buffer->data())
+        , m_end(m_begin + size())
+        , m_end_cap(m_begin + m_concrete_buffer->size())
+    {
+    }
+
+    BufferExpander(size_type nbyte, ctor_passkey const &)
+    {
+        expand(nbyte);
+    }
+
+    BufferExpander(ctor_passkey const &) {}
+
+    BufferExpander() = delete;
+    BufferExpander(BufferExpander const &) = delete;
+    BufferExpander(BufferExpander &&) = delete;
+    BufferExpander & operator=(BufferExpander const &) = delete;
+    BufferExpander & operator=(BufferExpander &&) = delete;
+    ~BufferExpander() = default;
+
+    size_type size() const noexcept
+    {
+        return static_cast<size_type>(this->m_end - this->m_begin);
+    }
+
+    size_type capacity() const noexcept
+    {
+        return static_cast<size_type>(this->m_end_cap - this->m_begin);
+    }
+
+    void reserve(size_type cap);
+
+    void expand(size_type length)
+    {
+        reserve(length);
+        m_end = m_begin + length;
+    }
+
+    std::shared_ptr<ConcreteBuffer> copy_concrete(size_type cap = 0) const;
+    std::shared_ptr<ConcreteBuffer> const & as_concrete(size_type cap = 0);
+    bool is_concrete() const { return bool(m_concrete_buffer); }
+
+    int8_t operator[](size_t it) const { return data(it); }
+    int8_t & operator[](size_t it) { return data(it); }
+
+    int8_t at(size_t it) const
+    {
+        validate_range(it);
+        return data(it);
+    }
+    int8_t & at(size_t it)
+    {
+        validate_range(it);
+        return data(it);
+    }
+
+    /* Backdoor */
+    int8_t data(size_type it) const { return data()[it]; }
+    int8_t & data(size_type it) { return data()[it]; }
+    int8_t const * data() const noexcept { return data<int8_t>(); }
+    int8_t * data() noexcept { return data<int8_t>(); }
+    // clang-format off
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+    template <typename T> T const * data() const noexcept { return reinterpret_cast<T *>(m_begin); }
```

Comment:
maybe we can have `assert` here to ensure the size is the multiple of the size of `T`?

Reply:
It may not be a good idea to add `assert` at a helper that should be performant and optimized away.

Are you talking about `m_begin`?  Having alignment check sounds like a good idea and it should be enforced all the time, not just `assert` (at debugging).  The allocation helper is the place to add the enforcement.  It may be tracked in a new issue.

Reply:
@yungyuc  Yes, `m_begin` should be checked. I agree with tracking this with a new issue.

Reply:
Feel free to create an issue for the alignment support in `ConcreteBuffer` and `BufferExpander`.  The alignment support should not use template argument because I want the two classes to remain to be classes.